File: agenkit-zig/tests/test_llm_validation.zig
/// Tests for LLM parameter validation

const std = @import("std");
const testing = std.testing;
const llm = @import("agenkit").adapter.llm;

test "CallOptions.withTemperature - valid values" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    // Valid boundary values
    try options.withTemperature(0.0);
    try testing.expectEqual(@as(f64, 0.0), options.temperature.?);

    try options.withTemperature(1.0);
    try testing.expectEqual(@as(f64, 1.0), options.temperature.?);

    try options.withTemperature(2.0);
    try testing.expectEqual(@as(f64, 2.0), options.temperature.?);
}

test "CallOptions.withTemperature - invalid too low" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    const result = options.withTemperature(-0.1);
    try testing.expectError(error.InvalidTemperature, result);
}

test "CallOptions.withTemperature - invalid too high" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    const result = options.withTemperature(2.1);
    try testing.expectError(error.InvalidTemperature, result);
}

test "CallOptions.withTemperature - invalid very negative" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    const result = options.withTemperature(-100.0);
    try testing.expectError(error.InvalidTemperature, result);
}

test "CallOptions.withMaxTokens - valid values" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    // Valid values
    try options.withMaxTokens(1);
    try testing.expectEqual(@as(usize, 1), options.max_tokens.?);

    try options.withMaxTokens(100);
    try testing.expectEqual(@as(usize, 100), options.max_tokens.?);

    try options.withMaxTokens(4096);
    try testing.expectEqual(@as(usize, 4096), options.max_tokens.?);
}

test "CallOptions.withMaxTokens - invalid zero" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    const result = options.withMaxTokens(0);
    try testing.expectError(error.InvalidMaxTokens, result);
}

test "CallOptions.withTopP - valid values" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    // Valid boundary values
    try options.withTopP(0.0);
    try testing.expectEqual(@as(f64, 0.0), options.top_p.?);

    try options.withTopP(0.5);
    try testing.expectEqual(@as(f64, 0.5), options.top_p.?);

    try options.withTopP(1.0);
    try testing.expectEqual(@as(f64, 1.0), options.top_p.?);
}

test "CallOptions.withTopP - invalid too low" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    const result = options.withTopP(-0.1);
    try testing.expectError(error.InvalidTopP, result);
}

test "CallOptions.withTopP - invalid too high" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    const result = options.withTopP(1.1);
    try testing.expectError(error.InvalidTopP, result);
}

test "CallOptions - multiple valid parameters" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    // Set all valid parameters
    try options.withTemperature(0.7);
    try options.withMaxTokens(1024);
    try options.withTopP(0.9);

    // Verify all values are set correctly
    try testing.expectEqual(@as(f64, 0.7), options.temperature.?);
    try testing.expectEqual(@as(usize, 1024), options.max_tokens.?);
    try testing.expectEqual(@as(f64, 0.9), options.top_p.?);
}

test "CallOptions - one invalid parameter should fail" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    // Valid parameters
    try options.withMaxTokens(100);
    try options.withTopP(0.9);

    // Invalid parameter should fail
    const result = options.withTemperature(3.0);
    try testing.expectError(error.InvalidTemperature, result);
}

test "CallOptions - boundary values" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    // Temperature boundaries
    try options.withTemperature(0.0);
    try testing.expectEqual(@as(f64, 0.0), options.temperature.?);
    try options.withTemperature(2.0);
    try testing.expectEqual(@as(f64, 2.0), options.temperature.?);

    // Max tokens boundary
    try options.withMaxTokens(1);
    try testing.expectEqual(@as(usize, 1), options.max_tokens.?);

    // Top P boundaries
    try options.withTopP(0.0);
    try testing.expectEqual(@as(f64, 0.0), options.top_p.?);
    try options.withTopP(1.0);
    try testing.expectEqual(@as(f64, 1.0), options.top_p.?);
}

test "CallOptions - undefined parameters are allowed" {
    const allocator = testing.allocator;
    var options = llm.CallOptions.init(allocator);
    defer options.deinit();

    // Initially all parameters should be null
    try testing.expect(options.temperature == null);
    try testing.expect(options.max_tokens == null);
    try testing.expect(options.top_p == null);

    // Set one parameter
    try options.withTemperature(0.7);
    try testing.expectEqual(@as(f64, 0.7), options.temperature.?);

    // Others should still be null
    try testing.expect(options.max_tokens == null);
    try testing.expect(options.top_p == null);
}

File: agenkit/tests/test_llm_validation.py
"""Test LLM parameter validation across all adapters."""

import pytest

from agenkit.adapters.llm.base import LLM


class MockLLM(LLM):
    """Mock LLM for testing validation."""

    async def complete(self, messages, **kwargs):
        """Mock complete method."""
        return None

    async def stream(self, messages, **kwargs):
        """Mock stream method."""
        yield None

    @property
    def model(self):
        """Mock model property."""
        return "mock-model"


class TestLLMValidation:
    """Test parameter validation for LLM adapters."""

    def test_temperature_validation(self):
        """Temperature must be between 0 and 2."""
        llm = MockLLM()

        # Valid temperatures
        llm._validate_llm_params(temperature=0.0)
        llm._validate_llm_params(temperature=1.0)
        llm._validate_llm_params(temperature=2.0)

        # Invalid: too low
        with pytest.raises(ValueError, match="temperature must be between 0 and 2"):
            llm._validate_llm_params(temperature=-0.1)

        # Invalid: too high
        with pytest.raises(ValueError, match="temperature must be between 0 and 2"):
            llm._validate_llm_params(temperature=2.1)

        # Invalid: wrong type
        with pytest.raises(ValueError, match="temperature must be a number"):
            llm._validate_llm_params(temperature="1.0")

    def test_max_tokens_validation(self):
        """Max tokens must be a positive integer."""
        llm = MockLLM()

        # Valid max_tokens
        llm._validate_llm_params(max_tokens=1)
        llm._validate_llm_params(max_tokens=100)
        llm._validate_llm_params(max_tokens=4096)

        # Invalid: zero
        with pytest.raises(ValueError, match="max_tokens must be positive"):
            llm._validate_llm_params(max_tokens=0)

        # Invalid: negative
        with pytest.raises(ValueError, match="max_tokens must be positive"):
            llm._validate_llm_params(max_tokens=-100)

        # Invalid: wrong type
        with pytest.raises(ValueError, match="max_tokens must be an integer"):
            llm._validate_llm_params(max_tokens=100.5)

    def test_top_p_validation(self):
        """Top_p must be between 0 and 1."""
        llm = MockLLM()

        # Valid top_p
        llm._validate_llm_params(top_p=0.0)
        llm._validate_llm_params(top_p=0.5)
        llm._validate_llm_params(top_p=1.0)

        # Invalid: too low
        with pytest.raises(ValueError, match="top_p must be between 0 and 1"):
            llm._validate_llm_params(top_p=-0.1)

        # Invalid: too high
        with pytest.raises(ValueError, match="top_p must be between 0 and 1"):
            llm._validate_llm_params(top_p=1.1)

    def test_frequency_penalty_validation(self):
        """Frequency penalty must be between -2 and 2."""
        llm = MockLLM()

        # Valid frequency_penalty
        llm._validate_llm_params(frequency_penalty=-2.0)
        llm._validate_llm_params(frequency_penalty=0.0)
        llm._validate_llm_params(frequency_penalty=2.0)

        # Invalid: too low
        with pytest.raises(ValueError, match="frequency_penalty must be between -2 and 2"):
            llm._validate_llm_params(frequency_penalty=-2.1)

        # Invalid: too high
        with pytest.raises(ValueError, match="frequency_penalty must be between -2 and 2"):
            llm._validate_llm_params(frequency_penalty=2.1)

    def test_presence_penalty_validation(self):
        """Presence penalty must be between -2 and 2."""
        llm = MockLLM()

        # Valid presence_penalty
        llm._validate_llm_params(presence_penalty=-2.0)
        llm._validate_llm_params(presence_penalty=0.0)
        llm._validate_llm_params(presence_penalty=2.0)

        # Invalid: too low
        with pytest.raises(ValueError, match="presence_penalty must be between -2 and 2"):
            llm._validate_llm_params(presence_penalty=-2.1)

        # Invalid: too high
        with pytest.raises(ValueError, match="presence_penalty must be between -2 and 2"):
            llm._validate_llm_params(presence_penalty=2.1)

    def test_multiple_parameters(self):
        """Validate multiple parameters at once."""
        llm = MockLLM()

        # All valid
        llm._validate_llm_params(
            temperature=0.7,
            max_tokens=100,
            top_p=0.9,
            frequency_penalty=0.5,
            presence_penalty=-0.5,
        )

        # One invalid parameter should raise
        with pytest.raises(ValueError, match="temperature"):
            llm._validate_llm_params(
                temperature=3.0,  # Invalid
                max_tokens=100,
                top_p=0.9,
            )
